DBZ-5283 Introduce a new SMT ExtractChangedRecordState to capture the names of changed/unchanged field within headers (#3980)
debezium/debezium: documentation/modules/ROOT/pages/transformations/event-flattening.adoc

When the SMT adds metadata fields to the simplified record's header, it prefixes each metadata field name with a double underscore. For a struct specification, the SMT also inserts an underscore between the struct name and the field name. +
 +
If you specify a field that is not in the change event record, the SMT does not add the field to the header.

[id="extract-new-record-state-drop-fields-header-name"]

> |xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`]

Reviewer: ```suggestion
|xref:extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`]
```

|

> |The Kafka message header name to source fields that should be dropped.

Reviewer: ```suggestion
|The Kafka message header name to use for listing field names in the source message that you want to drop from the output message.
```
Does this specify the name of  the header that you want to use to list the fields in the source message that you want to drop in the output message? Or does it represent the name(s) of the headers in the source that you want to drop from the output message?

Reviewer 2: The fields, so I believe your suggestion is correct.

[id="extract-new-record-state-drop-fields-from-key"]

> |xref:{link-event-flattening}#extract-new-record-state-drop-fields-from-key[`drop.fields.from.key`]

Reviewer: ```suggestion
|xref:extract-new-record-state-drop-fields-from-key[`drop.fields.from.key`]
```

|`false`

> |Controls whether fields are also be dropped fom the event's key.

Reviewer: ```suggestion
|Specifies whether you want the SMT to remove fields that are listed in `drop.fields.header.name` from the event's key.
```
I'm assuming that the fields to remove from the key are those that you specify in `drop.fields.header.name`?

Reviewer 2: Correct.

[id="extract-new-record-state-drop-fields-keep-schema-compatible"]

> |xref:{link-event-flattening}#extract-new-record-state-drop-fields-keep-schema-compatible[`drop.fields.keep.schema.compatible`]

Reviewer: ```suggestion
|xref:extract-new-record-state-drop-fields-keep-schema-compatible[`drop.fields.keep.schema.compatible`]
```

|`true`

> |Controls whether non-optional fields will be removed if they're included in the xref:{link-event-flattening}#extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`] configuration property. +

Reviewer: ```suggestion
|Specifies whether you want the SMT to remove non-optional fields that are included in the xref:extract-new-record-state-drop-fields-header-name[`drop.fields.header.name`] configuration property. +
```

 +

> The default behavior is to only remove fields if they're marked optional.

Reviewer: ```suggestion
By default, the SMT only removes fields that are marked `optional`.
```

 +
If you specify a field that is not in the change event record, the SMT does not add the field to the header.
|===

The following table describes the options that you can specify to configure the event flattening SMT to integrate with the event changes SMT.

.Descriptions of event flattening SMT configuration options for integration with event changes SMT
[cols="30%a,25%a,45%a",subs="+attributes",options="header"]
|===
|Option
|Default
|Description

> |[[extract-new-record-state-header-changed-name]]xref:{link-event-flattening}#extract-new-record-state-header-unchanged-name[`header.unchanged.name`]

Reviewer: ```suggestion
|[[extract-new-record-state-header-changed-name]]xref:extract-new-record-state-header-unchanged-name[`header.unchanged.name`]
```
There's no need to include the category attribute for xrefs that link to content in the same file. Same comment applies to the table entries that follow.

|
|The Kafka message header to source unchanged field names from.

[id="extract-new-record-state-drop-unchanged-fields"]

> |xref:{link-event-flattening}#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]

Reviewer: ```suggestion
|xref:extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`]
```

|`false`
|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.

> For an `UPDATE` operation, this behavior can be adjusted to only include the fields that were changed by setting this option to `true` to drop or omit fields that were not changed in the update.

Reviewer: ```suggestion
For an `UPDATE` operation, set this option to `true` if you want the SMT to only include fields that were changed during an operation. Fields that were not changed in the update are dropped. 
```

[id="extract-new-record-state-keep-unchanged-primary-keys"]

> |xref:{link-event-flattening}#extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]

Reviewer: ```suggestion
|xref:extract-new-record-state-keep-unchanged-primary-keys[`keep.unchanged.primary.keys`]
```

|`true`
|{prodname} generates a flattened event structure based on the `after` portion of the event's payload that includes all fields by default.

> When enabling link:#extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`], it is possible to omit primary key values in the flattened structure by setting this option to `false`.

Reviewer: ```suggestion
If xref:extract-new-record-state-drop-unchanged-fields[`drop.unchanged.fields`] is set to `true`, set `keep.unchanged.primary.keys` to `false` to omit primary key values from the flattened structure.
```

> By default, primary key fields are always retained even if unchanged fields are to be omitted.

Reviewer: ```suggestion
The default setting (`true`) retains primary key fields even if the event changes SMT (`ExtractChangedRecordState`) is configured to omit unchanged fields.
```
